specfile/macro_options.py

```diff
@@ -0,0 +1,523 @@
+        option = f"-{name}"
+        for i, token in reversed(list(enumerate(self._tokens))):
+            if not token.value.startswith(option):
+                continue
+            if token.value != option:
+                return i, i
+            if not self._requires_argument(name):
+                return i, -1
+            j = i + 1
+            if j == len(self._tokens):
+                return i, -1
+            if self._tokens[j].type == TokenType.WHITESPACE:
+                j += 1
+            if j == len(self._tokens):
+                return i, -1
+            if self._tokens[j].value.startswith("-"):
+                return i, -1
+            return i, j
+        return -1, -1
+
+    @staticmethod
+    def _needs_quoting(value):
+        # if there is a whitespace, enquote the value rather than escaping it
+        return any(ws in str(value) for ws in string.whitespace)
+
+    def __getattr__(self, name: str) -> Union[bool, int, str]:
+        if not self._valid_option(name):
+            return super().__getattribute__(name)
+        i, j = self._find_option(name)
+        if i < 0:
+            if self._requires_argument(name):
+                return self.defaults.get(name)
+            return False
+        value = self._tokens[j].value if j > i else self._tokens[i].value[2:]
+        if not value:
+            return True
+        if value.isnumeric():
+            return int(value)
+        return value
+
+    def __setattr__(self, name: str, value: Union[bool, int, str]) -> None:
+        if not self._valid_option(name):
+            return super().__setattr__(name, value)
+        if self._requires_argument(name) and isinstance(value, bool):
+            raise MacroOptionsException(f"Option -{name} requires an argument.")
+        if (
+            not self._requires_argument(name)
+            and not isinstance(value, bool)
+            and value is not None
+        ):
+            raise MacroOptionsException(f"Option -{name} is a flag.")
+        i, j = self._find_option(name)
+        if i < 0 and value is not None and value is not False:
+            if self._tokens:
+                self._tokens.append(Token(TokenType.WHITESPACE, " "))
+            if value is True:
+                self._tokens.append(Token(TokenType.DEFAULT, f"-{name}"))
+            elif isinstance(value, int):
+                self._tokens.append(Token(TokenType.DEFAULT, f"-{name}{value}"))
+            else:
+                self._tokens.append(Token(TokenType.DEFAULT, f"-{name}"))
+                self._tokens.append(Token(TokenType.WHITESPACE, " "))
+                self._tokens.append(
+                    Token(
+                        TokenType.DOUBLE_QUOTED
+                        if self._needs_quoting(value)
+                        else TokenType.DEFAULT,
+                        value,
+                    )
+                )
+            return
+        if value is None or value is False:
+            return delattr(self, name)
```

Comment:
I may be missing something here, could you please explain, why is `__setattr__` deleting?

Reply:
If you set a flag option to `False`, it is effectively removed:

```python
> str(prep.setup)
"%setup -q"
> prep.setup.options.q = False
> str(prep.setup)
"%setup"
```

And setting an option to `None` seemed like a convenient alternative way to remove it (isn't it better than raising an exception?).

Reply:
Thanks for the explanation, sounds good